File: 链表/24.两两交换链表中的节点.java
/**
 * @author 吴洋
 */
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) {this.val = val; this.next = next;}
}
/*
 * @lc app=leetcode.cn id=24 lang=java
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution24 {
    // public ListNode swapPairs(ListNode head) {
    //     ListNode dummyHead = new ListNode(0, head);
    //     ListNode cur = dummyHead;
    //     while(cur.next != null && cur.next.next != null) {
    //         ListNode temp1 = cur.next;
    //         ListNode temp2 = cur.next.next.next;
    //         cur.next = cur.next.next;
    //         cur.next.next = temp1;
    //         cur.next.next.next = temp2;
    //         cur = cur.next.next;
    //     }
    //     return dummyHead.next;
    // }

    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null) {
            return head;
        }
        ListNode newHead = head.next;
        head.next = swapPairs(newHead.next);
        newHead.next = head;
        return newHead;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
        Solution24 solution = new Solution24();
        ListNode result = solution.swapPairs(head);
        while(result != null) {
            System.out.println(result.val);
            result = result.next;
        }
    }
}
// @lc code=end

File: 链表/24.两两交换链表中的节点.ts

class ListNode2 {
    val: number;
    next: ListNode2 | null;
    constructor(val?: number, next?: ListNode2 | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}
/*
 * @lc app=leetcode.cn id=24 lang=typescript
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function swapPairs1(head: ListNode2 | null): ListNode2 | null {
    let dummyHead = new ListNode2(0, head);
    let cur = dummyHead;
    while(cur.next != null && cur.next.next != null) {
        let temp1 = cur.next;
        let temp2 = cur.next.next.next;
        cur.next = cur.next.next;
        cur.next.next = temp1;
        cur.next.next.next = temp2;
        cur = cur.next.next;
    }
    return dummyHead.next;
};
// @lc code=end

